### DISCUSSION TITLE
Example repo on GitLab using PDM and nox

### DISCUSSION THREAD

**Discussion by pohlt** [pdm-project/pdm#327] [Q&A]
Hi All,
I set up an [example repo on GitLab using PDM and nox](https://gitlab.com/coherentminds/pdm-gitlab-example).

Could someone please have a look if the setup makes sense? In particular, I'm not sure about the following points:
* How to install nox? Currently, it's installed globally using pip. Should it also be handled by pdm?
* How to handle different versions of required packages using nox (`devtools` in this example)? Is it really optimal to have this two step approach with pdm (see [test step in the noxfile](https://gitlab.com/coherentminds/pdm-gitlab-example/-/blob/main/noxfile.py))?

Please feel free to comment on any other topic, since I'm rather new to CI/CD on GitLab and eager to learn from pros.

Cheers,
Tom

**Comment by frostming** [comment]
> How to install nox? Currently, it's installed globally using pip. Should it also be handled by pdm?

Either way is Okay, defining it in dev dependencies makes the project battery included and developers need little setup.
Installing it globally, by `pipx` is the way I prefer. This reduces duplicated dependencies shared by multiple projects.

> How to handle different versions of required packages using nox (devtools in this example)? Is it really optimal to have this two step approach with pdm (see test step in the noxfile)?

I would recommend using `pip` directly because `pdm add` when running test will modify the pyproject.toml file and cost more time locking deps.
